dirsrvtests/tests/perf/ndncache_test.py

```diff
@@ -0,0 +1,577 @@
+from lib389.config import BDB_LDBMConfig, LMDB_LDBMConfig
+from lib389.config import Config
+from lib389.index import Indexes
+from lib389._mapped_object import DSLdapObject
+from lib389.monitor import MonitorBackend
+from lib389.properties import TASK_WAIT
+from lib389.topologies import topology_st as topo
+from shutil import copyfile
+from statistics import fmean, stdev
+
+
+from lib389._constants import (
+    DEFAULT_BENAME,
+    DEFAULT_SUFFIX,
+    DN_DM,
+    DN_USERROOT_LDBM,
+    PASSWORD,
+)
+
+from lib389.utils import (
+    ldap, os, time, logging,
+    ds_is_older,
+    ensure_bytes,
+    ensure_str,
+    get_default_db_lib,
+    get_ldapurl_from_serverid,
+)
+
+pytestmark = pytest.mark.tier3
+
+THIS_DIR = os.path.dirname(__file__)
+LDIF = os.path.join(THIS_DIR, '../data/5Kusers.ldif')
+RESULT_DIR = f'{THIS_DIR}/../data/ndncache_test_results/r'
+RESULT_FILE = f'{RESULT_DIR}/results_ndncache.'
+CSV_FILE = f'{RESULT_DIR}/r.csv.'
+NB_MEASURES = 100
+NB_MEANINGFULL_MEASURES = NB_MEASURES-80
+SCENARIO='SCENARIO'
+STAT_ATTRS2 =  ( 'currentdncachesize', 'currentdncachecount' )
+STAT_ATTRS = ( 'dncachehits', 'dncachetries' ) + STAT_ATTRS2
+
+WITHOUT_CACHE = 'without_ndn_cache'
+WITH_CACHE = 'with_ndn_cache'
+
+
+logging.basicConfig(level=logging.DEBUG)
+log = logging.getLogger(__name__)
+
+class Scenario:
+    def __init__(self):
+        self.ldc = None
+        self.results = {}
+        self._desc = None
+        self._name = None
+
+    def preop(self):
+        # Run before measured operation
+        pass
+
+    @abstractmethod
+    def op(self):
+        # Operation to measure
+        pass
+
+    def postop(self):
+        # Run after measured operation
+        pass
+
+    def __str__(self):
+        return self._name
+
+    def description(self):
+        return self._desc
+
+    @staticmethod
+    def average(data):
+        data.sort()
+        log.info(f'RAW DATA: len(data)={len(data)} {data}')
+        assert len(data) >= NB_MEANINGFULL_MEASURES
+        for i in range(len(data)-1, NB_MEANINGFULL_MEASURES-1, -1):
+            del data[i]
+        log.info(f'COOKED DATA: len(data)={len(data)} {data}')
+        m = fmean(data)
+        v = stdev(data, m)
+        v = v/m * 100
+        return [ m, v, str(data) ]
+
+    def getndncache_stats(self):
+        dnbemon = 'cn=monitor,cn=userRoot,cn=ldbm database,cn=plugins,cn=config'
+
+        """The following fails ...
+        mon = MonitorBackend(inst, dn=dnbemon)
+        res = mon.get_status()
+        log.debug(f'getndncache_stats res={res}')
+        return [ res[a] for a in STAT_ATTRS ]
+        """
+        res = self.ldc.search_ext_s(
+            base=dnbemon,
+            scope=ldap.SCOPE_BASE,
+            filterstr='(objectclass=*)',
+            attrlist=list(STAT_ATTRS))
+        log.debug(f'getndncache_stats res={res}')
+        res = res[0][1]
+        return { a: ensure_str(res[a][0]) for a in STAT_ATTRS }
+
+
+
+    def measure(self, inst, conn):
```

Comment:
This function takes forever to execute on a VM. I understand that we might focus on a faster machine execution, but if we can optimize it, it won't hurt (I can't realistically test it on any moderate VM now)